syz-manager/manager.go

```diff
@@ -668,6 +670,14 @@ func (mgr *Manager) saveCrash(crash *Crash) bool {
 		mgr.dataRaceFrames[crash.Frame] = true
 		mgr.mu.Unlock()
 	}
+	if crash.Type == report.MultiRead {
+		mgr.mu.Lock()
+		if mgr.multiReadFrames[crash.Title] {
+			mgr.cfg.Ignores = append(mgr.cfg.Ignores, crash.Title)
```

Comment:
Isn't there a race with reading Ignores?

Reply:
Yeah i think so i will try to solve it or to find another way to ignore duplicated reports

Reply:
@dvyukov do you think it is viable to lock mgr.cfg.Ignores when reading it as well to avoid this data race?
Would it affect the manager performance?

Reply:
I should be totally fine. These are read once per VM boot/run, which is way more expensive.

Reply:
by this as far as i understand, sending a mutex variable is needed as extra parameter in newReporter function? so is it a good idea?

Also from my understanding newReporter is called in the beginning of runManager before even vmloop(), is that mean that this function wont have the updated Ignore list when we modifying it everytime in saveCrash() function if crash type is multiRead?

Reply:
You are right.
Passing in mutex feels awkward. And, yes, we also need to re-read ignores somehow when we start a VM.

We also have this use of mgr.cfg:
https://github.com/google/syzkaller/blob/master/syz-manager/html.go#L77
and maybe more.

I think modifying the user config is the root cause of all the problems we have. It also has high chances of provoking bugs in future. I think we should leave the config read-only and collect the additional dynamic ignores elsewhere.

For NewReporter we could create a new reporter for each VM and then arrange passing in the additional dynamic ignores. But I am thinking maybe we should introduce notion of "dynamic ignores" explicitly and pass in e.g. a callback that returns additional dynamic ignores `ignores func() []*regexp.Regexp`? The manager implementation would use the mutex to guard reading of the ignores. Then we don't need to create reporter per VM and ignores will take effect immediately.

But I afraid there is another problem related to pkg/repro.Run. We pass in a reporter and it will need careful handling either way as this reporter must not ignore the bug we are trying to reproduce...

Perhaps we need to start with collecting and analyzing all uses of config and reporter.

Reply:
i think yeah collecting the uses of config and reporter will help us in finding the right way to ignore duplicated reports.
okay i can start doing that.

i have a last question, is saveCrash() returning false still leading to restart the instance or affect the performance ?
i mean is there a way to manage duplicated reports using this function or it wont help?

Reply:
> i have a last question, is saveCrash() returning false still leading to restart the instance or affect the performance ?
> i mean is there a way to manage duplicated reports using this function or it wont help?

No, it will not help, by that time the VM is already destroyed.

Reply:
I think we dont need the Ignores passed by config in manager, as much as we need the ignores ( the regex of Ignores) that is used by containsCrash() and used everytime crash is detected so i think we need somehow to make it accessible, i dont know if this will slow down the system or not as everytime we have a report we need to lock and check, but i think it is the only place that used everytime when a crash found.
is there a way to have benefits from this ignores list in duplicated reports, for example we can make the ignores contain the regex of Ignores as a start and then we can add to it regex of any crash we found for multiread ?

Reply:
> i dont know if this will slow down the system or not as everytime we have a report we need to lock and check, but i think it is the only place that used everytime a crash found.

I think it should be fine. In most cases a report means a kernel bug and VM recreation with all associated overheads. Moreover, we will need to lock the mutex for a very brief period only (to read the slice).

> is there a way to have benefits from this ignores list in duplicated reports, for example we can make the ignores contain the regex of Ignores as a start and then we can add to it regex of any crash we found for multiread ?

I don't understand the question. Please rephrase.

Reply:
i mean we need to modify this config https://github.com/google/syzkaller/blob/d46bc75207fea1d7671c1277dd660cf1a4d7847b/pkg/report/report.go#L107 to have a lock for ignores, and also need to check for a way  to modify ignores list everytime we detect a Bug before reporting it as crash and restarting the instance.
do you see it a good idea to prevent duplicated report using this way ? 
I check this config is used by all os reports like (pkg/report/linux.go, pkg/report/netbsd.go). so as far as i understand i dont think this will lead to a lot of modification:
 1) we just need to add extra element in this config for a lock.
2) find a good place that we can add regex of multiRead report to this ignore list.
3) its already check in matchOops() function if the crash not matched any of ignores, so no modification in this part.

Reply:
This config is an internal implementation detail of pkg/report (unexported type). Nobody outside of pkg/report should have access to it nor be bothered with it.
We still can achieve the end result of modifying the ignore list, but it need to happen by changing the public  pkg/report.NewReporter interface in some way. So the question is: how do we change the NewReporter interface?
One possible way would be what I proposed above with a callback.
But we also need some plan for pkg/repro problem I mentioned above.

Reply:
What if we make the dynamic ignores list a separate property of report.NewReporter and add a method that updates that list and recompiles the regexps from the builtin cfg ignores and from the dynamic list while holding a mutex that resides in NewReporter?
Regarding the pkg/repro.Run problem - we could probably maintain two lists of regexps (builtin ignores only / builtin + dynamic ignores) and choose between them when calling ContainsCrash and such?

Reply:
> What if we make the dynamic ignores list a separate property of report.NewReporter and add a method that updates that list and recompiles the regexps from the builtin cfg ignores and from the dynamic list while holding a mutex that resides in NewReporter?

Right, we don't have to store this list externally. A method that adds a new ignore is simpler than a callback.

> Regarding the pkg/repro.Run problem - we could probably maintain two lists of regexps (builtin ignores only / builtin + dynamic ignores) and choose between them when calling ContainsCrash and such?

I am thinking if complicating Reporter interface worth it, or we just create a new Reporter from scratch (with empty dynamic ignores). We could either create a new one per reproduction procedure, or create 2 persistent in the manager -- one with dynamic ignores, another -- without dynamic ignores. Reporter is a bunch of compiled regexps, which is not too much, an for linux it holds map of all kernel symbols/addresses... I can't make my mind what's better...